notebook/converted/visualize_errors.py

```diff
@@ -0,0 +1,848 @@
+
+# ### Are there any hints in the metadata about why the samples are hard to classify?
+
+# In[111]:
+
+
+metadata_df = metadata_df.set_index('external_id')
+
+
+# In[112]:
+
+
+metadata_df = metadata_df[metadata_df.index.notnull()]
+
+
+# In[113]:
+
+
+hard_sample_df = metadata_df.loc[hard_samples]
+hard_sample_df
+
+
+# In[114]:
+
+
+metadata_df['recount_seq_qc.avgq'] = pd.to_numeric(metadata_df['recount_seq_qc.avgq'], errors='coerce')
+metadata_df = metadata_df[metadata_df['recount_seq_qc.avgq'].notnull()]
+metadata_df
+
+
+# In[115]:
+
+
+metadata_df['recount_seq_qc.avgq'] = metadata_df['recount_seq_qc.avgq'].astype(float)
+metadata_df['recount_seq_qc.avgq'].plot.hist(title='AvgQ Distributon')
+
+
+# In[116]:
+
+
+hard_sample_df['recount_seq_qc.avgq'] = hard_sample_df['recount_seq_qc.avgq'].astype(float)
+hard_sample_df['recount_seq_qc.avgq'].plot.hist(title='AvgQ Distributon in Hard Samples')
+
+
+# ## Do the hard samples stand out in a UMAP embedding?
+
+# In[117]:
+
+
+expression_df = pd.read_pickle("../../data/no_scrna_tpm.pkl")
+
+
+# In[118]:
+
+
+# Running UMAP on 200k samples just makes a giant oval
+is_hard = [sample in hard_samples for sample in expression_df.index]
+expression_df['is_hard'] = is_hard
+
+easy_df = expression_df[expression_df['is_hard'] == False]
+easy_df = easy_df.sample(n=4000, random_state=42)
+hard_df = expression_df[expression_df['is_hard']]
+sampled_df = pd.concat([easy_df, hard_df])
+
+# Remove column to extract expression for UMAP
+is_hard = sampled_df['is_hard']
+sampled_df = sampled_df.drop(['is_hard'], axis='columns')
+
+sampled_df
+
+
+# In[119]:
+
+
+expression_matrix = sampled_df.to_numpy()
+expression_matrix.shape
+
+
+# In[120]:
+
+
+reducer = umap.UMAP(transform_seed=42, random_state=42, n_components=2)
+
+
+# In[121]:
+
+
+get_ipython().run_cell_magic('time', '', 'expression_umap = reducer.fit_transform(expression_matrix)')
+
+
+# In[122]:
+
+
+umap_df = pd.DataFrame(expression_umap, index=sampled_df.index, columns=['UMAP1', 'UMAP2'])
+
+
+# In[123]:
+
+
+umap_df
+
+
+# In[124]:
+
+
+umap_df['is_hard'] = is_hard
+
+
+# In[125]:
+
+
+umap_df
+
+
+# In[126]:
+
+
+plot = ggplot(umap_df, aes(x='UMAP1', y='UMAP2', fill='is_hard'))
+plot += geom_point()
+plot += ggtitle('UMAP Embedding of all samples')
+print(plot)
+
+
+# ## There seems to be a reasonable high-d linear dividing line since LR works well. Maybe PCA makes more sense for visualizing?
+
+# In[127]:
+
+
+reducer = PCA(random_state=42, n_components=2)
+
+
+# In[128]:
+
+
+get_ipython().run_cell_magic('time', '', 'expression_umap = reducer.fit_transform(expression_matrix)')
+
+
+# In[129]:
+
+
+pca_df = pd.DataFrame(expression_umap, index=sampled_df.index, columns=['UMAP1', 'UMAP2'])
+
+
+# In[130]:
+
+
+pca_df['is_hard'] = is_hard
+
+
+# In[131]:
+
+
+plot = ggplot(pca_df, aes(x='UMAP1', y='UMAP2', fill='is_hard'))
+plot += geom_point()
+plot += ggtitle('UMAP Embedding of all samples')
+print(plot)
```

Comment:
This is actually plotting the PCA embedding, right? Maybe you copied over your previous plotting code and forgot to change the embedding name.

Reply:
Good catch! Fixed it and some other umap -> pca swaps